docassemble/AssemblyLine/al_document.py

```diff
@@ -234,43 +238,43 @@ def safe_value(
+        # If we preserve newlines, we need to account for max_lines, not just max_chars
         if preserve_newlines and max_lines > 1:
             if isinstance(value, str):
+                max_lines = self.max_lines(
+                    input_width=input_width, overflow_message_length=0
+                )
+                max_chars = self.overflow_trigger
                 # Replace all new line characters with just \n. \r\n inserts two lines in a PDF
                 value = re.sub(r"[\r\n]+|\r+|\n+", r"\n", value).rstrip()
-                line = 1
-                retval = ""
-                paras = value.split("\n")
-                para = 0
-                while line <= max_lines and para < len(paras):
-                    # add the whole paragraph if less than width of input
-                    if len(paras[para]) <= input_width:
-                        retval += paras[para] + "\n"
-                        line += 1
-                        para += 1
-                    else:
-                        # Keep taking the first input_width characters until we hit max_lines
-                        # or we finish the paragraph
-                        while line <= max_lines and len(paras[para]):
-                            retval += paras[para][:input_width]
-                            paras[para] = paras[para][input_width:]
-                            line += 1
-                        if not len(paras[para]):
-                            para += 1
-                            retval += "\n"
-                # TODO: check logic here to only add overflow message when we exceed length
-                if len(paras) > para:
-                    return (
-                        retval.rstrip() + overflow_message
-                    )  # remove trailing newline before adding overflow message
-                else:
-                    return retval
+                # textwrap.wrap does all the hard work for us here
+                return " ".join(
+                    wrap(
+                        value,
+                        width=input_width,
+                        max_lines=max_lines,
+                        replace_whitespace=False,
+                        placeholder=overflow_message,
+                    )
+                ).replace("  ", " ")
 
-        # Strip newlines from strings
+        # Strip newlines from strings because they take extra space
         if isinstance(value, str):
             if len(value) > self.overflow_trigger:
+                if preserve_words:
+                    return (
+                        next(
+                            iter(
+                                wrap(
+                                    value,
+                                    width=max_chars,
+                                    replace_whitespace=True,
+                                    drop_whitespace=True,
+                                )
+                            )
+                        )
```

Comment:
formatter decided to go a little crazy with this one 😅  Maybe break it into a few lines